### PR TITLE
Adds sanity checking to prefs checks, fixing a bug that can cause emotes to stop displaying intermittently

### PR DESCRIPTION
**Pull request by vinylspiders**
## About The Pull Request

This PR adds a bunch of sanity checking to the `prefs.chat_toggles` var accesses. 

Prefs can be null when client is creating/deleting and this causes a runtime where you might not be able to use emotes.

image

EDIT: as per suggestion I stopped masking the issue and added stack traces instead. And made them into helper procs to avoid code duplication.

For some of these procs it is important that we are able to continue running to get to the end. **A ghost having null prefs should not stop everyone else from hearing/seeing the emote.** 

So now they will put the stack trace and keep on trucking if that issue occurs. This is the main 'point' of the PR, which has seemingly gotten bogged down in the creation of the two helper procs--apologies. 

See https://github.com/tgstation/tgstation/pull/70404 for essentially the same issue but applied to chat messages. I have gone back and replaced the duplicated code from that to use the new helper procs instead.

## Why It's Good For The Game

Anything chat related is a bad place to runtime. Some messages can take a while to type out and to have it not display is frustrating at worst, possibly detrimental if you have to communicate something urgently.

## Changelog

:cl:
fix: fixes a bug that can cause emotes to stop working if a client is being created or deleted
/:cl:

### CODE REVIEW

**Review by ZephyrTFA — CHANGES_REQUESTED** [review]
This is hiding a larger problem. The prefs datum should never be null

File: code/datums/brain_damage/imaginary_friend.dm [hunk 1]
```diff
@@ -250,7 +252,10 @@
 	// Speech bubble, but only for those who have runechat off
 	var/list/speech_bubble_recipients = list()
 	for(var/mob/user as anything in (group + src)) // Add ourselves back in
-		if(user.client && (!user.client.prefs.read_preference(/datum/preference/toggle/enable_runechat) || (SSlag_switch.measures[DISABLE_RUNECHAT] && !HAS_TRAIT(src, TRAIT_BYPASS_MEASURES))))
+		if(user.client && !user.client?.prefs)
+			stack_trace("[usr] ([usr.ckey]) had null prefs, which shouldn't be possible!")
```
Comment (MrMelbert) [inline code/datums/brain_damage/imaginary_friend.dm:256]: `usr` is likely not the same as `user`

**Review by MrMelbert — COMMENTED** [review]
If we are to go through with this, surely we should have a wrapper for `mob/proc/safe_read_preference` that throws the stack trace or returns the pref rather than copy and pasting it to every place where it's needed

Though it makes me wonder if it's not possible, why is it happening so much? If it's like you said (prefs is getting null'd from the client being deleted) is there another way we can avoid this?

**Comment by optimumtact** [comment]
are yo usure it's because the prefs datum is null and not because the client datum is null?

**Comment by MrMelbert** [comment]
> @MrMelbert as for the cause--my suspicion is that it lies in the fact that we access it directly via `prefs.chat_toggles` rather than `read_preference()` (which I assume guards against null prefs?) Does that sound right? @optimumtact ?

`read_preference` doesn't guard against anything in regards to null prefs because it's a prefs proc

**Comment by Mothblocks** [comment]
I no longer understand the point of this PR, if we're going to be throwing a runtime and returning false anyway what is the point? These nulls already should be cancelling execution I would think? This just gives people the wrong idea when reading these consumers imo

**Review by Ghommie — APPROVED** [review]

**Comment by Ghommie** [comment]
A merge conflict has arisen. Sorry if I forgot to merge this before.

**Pull request merged**